homework/ba-nguyen/lesson-7.2/javascript-3.md

```diff
@@ -0,0 +1,249 @@
+
+  timeout = setInterval(() => {
+    if (from <= to) {
+      console.log(from);
+    }
+
+    if (from === to) {
+      clearInterval(timeout);
+    }
+
+    from++;
+  }, 1000);
+}
+```
+
+- Using nested setTimeout
+
+```typescript
+function printNumbers(from: number, to: number) {
+  if (from > to) return;
+
+  setTimeout(() => {
+    console.log(from);
+    printNumbers(from + 1, to);
+  }, 1000);
+}
+```
+
+**Question 2**: Use the **reduce** method in combination with the **concat** method to “flatten” an array of arrays into a single array that has all the elements of the original arrays.
+
+```typescript
+const arrays = [[1, 2, 3], [4, 5], [6]];
+→ [1, 2, 3, 4, 5, 6]
+
+function flatten(arrays: any[], depth: number = Infinity): any[] {
+  if (!Array.isArray(arrays)) {
+    return arrays;
+  }
+
+  if (depth > 0) {
+    return arrays.reduce(
+      (prevArr, curVal) => prevArr.concat(flatten(curVal, depth - 1)),
+      []
+    );
+  }
+
+  return arrays.slice();
+}
+```
+
+**Question 3**: Write a higher-order function loop that provides something like a for loop statement. It takes a value, a test function, an update function, and a body function. Each iteration, it first runs the test function on the current loop value and stops if that returns false. Then it calls the body function, giving it the current value. Finally, it calls the update function to create a new value and starts from the beginning. When defining the function, you can use a regular loop to do the actual looping.
+
+loop(3, n => n > 0, n => n - 1, console.log);
+// → 3
+// → 2
+// → 1
+
+```typescript
+type TestFuncType = (n: number) => boolean;
+type UpdateFuncType = (n: number) => number;
+type BodyFuncType = (n: number) => void;
+
+function loop(
+  value: number,
+  testFunc: TestFuncType,
+  updateFunc: UpdateFuncType,
+  bodyFunc: BodyFuncType
+) {
+  while (testFunc(value)) {
+    bodyFunc(value);
+    value = updateFunc(value);
+  }
+}
+```
+
+**Question 4**: Write a function, **deepEqual**, that takes two values and returns true only if they are the same value or are objects with the same properties whose values are also equal when compared with a recursive call to deepEqual.
+
+```typescript
+function deepEqual<T1, T2>(value1: T1 | T2, value2: T1 | T2) {
+  if (typeof value1 !== typeof value2) return false;
+
+  if (
+    typeof value1 === "object" &&
+    Array.isArray(value1) &&
+    Array.isArray(value2)
+  ) {
+    const newValue1 = value1.sort();
+    const newValue2 = value2.sort();
+
+    return (
+      value1.length === value2.length &&
+      newValue1.every((item, index) => deepEqual(item, newValue2[index]))
+    );
+  }
+
+  if (typeof value1 === "object") {
+    const keys1 = Object.keys(value1);
+    const keys2 = Object.keys(value2);
+
+    return (
+      keys1.length === keys2.length &&
+      keys1.every((item, index) => deepEqual(item, keys2[index]))
+    );
+  }
+
+  return value1 === value2;
+}
+
+console.log(deepEqual([1, 2], [1, 2]));
```

Comment:
some test case not pass

const obj = {here: {is: "an", other: "3"}, object: 2};
console.log(deepEqual(obj, obj));
// → true
console.log(deepEqual(obj, {here: 1, object: 2}));
// → false
console.log(deepEqual(obj, {here: {is: "an"}, object: 2}));
// → false
console.log(deepEqual(obj, {here: {is: "an", other: "2"}, object: 2}));
// → false
console.log(deepEqual(obj, {here: {is: "an", other: "3"}, object: 2}));
// → true

Reply:
I update my code and you can double check again